<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Remove chimeric molecules — chimericDrops • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Remove chimeric molecules — chimericDrops" />
<meta property="og:description" content="Remove chimeric molecules within each cell barcode's library in a droplet experiment." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Remove chimeric molecules</h1>
    
    <div class="hidden name"><code>chimericDrops.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Remove chimeric molecules within each cell barcode's library in a droplet experiment.</p>
    </div>

    <pre class="usage"><span class='fu'>chimericDrops</span><span class='op'>(</span><span class='va'>sample</span>, barcode.length <span class='op'>=</span> <span class='cn'>NULL</span>, use.library <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>removeChimericDrops</span><span class='op'>(</span>
  <span class='va'>cells</span>,
  <span class='va'>umis</span>,
  <span class='va'>genes</span>,
  <span class='va'>nreads</span>,
  <span class='va'>ref.genes</span>,
  min.frac <span class='op'>=</span> <span class='fl'>0.8</span>,
  get.chimeric <span class='op'>=</span> <span class='cn'>FALSE</span>,
  get.diagnostics <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sample</th>
      <td><p>String containing paths to the molecule information HDF5 files, 
produced by CellRanger for 10X Genomics data.</p></td>
    </tr>
    <tr>
      <th>barcode.length</th>
      <td><p>An integer scalar specifying the length of the cell barcode, see <code><a href='read10xMolInfo.html'>read10xMolInfo</a></code>.</p></td>
    </tr>
    <tr>
      <th>use.library</th>
      <td><p>An integer vector specifying the library indices for which to extract molecules from <code>sample</code>.
Alternatively, a character string specifying one or more library types, e.g., <code>"Gene expression"</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further arguments to be passed to <code>removeChimericDrops</code>.</p></td>
    </tr>
    <tr>
      <th>cells</th>
      <td><p>Character vector containing cell barcodes, where each entry corresponds to one molecule.</p></td>
    </tr>
    <tr>
      <th>umis</th>
      <td><p>Integer vector containing encoded UMI sequences, see <code>?<a href='encodeSequences.html'>encodeSequences</a></code> for details.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Integer vector specifying the gene indices.
Each index should refer to an element of <code>ref.genes</code>.</p></td>
    </tr>
    <tr>
      <th>nreads</th>
      <td><p>Integer vector containing the number of reads per molecule.</p></td>
    </tr>
    <tr>
      <th>ref.genes</th>
      <td><p>A character vector containing the names or symbols of all genes.</p></td>
    </tr>
    <tr>
      <th>min.frac</th>
      <td><p>A numeric scalar specifying the minimum fraction of reads required for a chimeric molecule to be retained.</p></td>
    </tr>
    <tr>
      <th>get.chimeric</th>
      <td><p>A logical scalar indicating whether the UMI counts corresponding to chimeric molecules should be returned.</p></td>
    </tr>
    <tr>
      <th>get.diagnostics</th>
      <td><p>A logical scalar indicating whether to return statistics for each molecule grouping.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list is returned with the <code>cleaned</code> entry, 
a sparse matrix containing the UMI count for each gene (row) and cell barcode (column) after removing chimeric molecules.
All cell barcodes that were originally observed are reported as columns, 
though note that it is theoretically possible for some barcodes to contain no counts.</p>
<p>If <code>get.chimeric=TRUE</code>, a <code>chimeric</code> entry is returned in the list.
This is a sparse matrix of UMI counts corresponding to the chimeric molecules.
Adding the cleaned and chimeric matrices should yield the total UMI count prior to removal of swapped molecules.</p>
<p>If <code>get.diagnostics=TRUE</code>, the top-level list will also contain an additional <code>diagnostics</code> DataFrame.
Each row corresponds to a group of molecules in the same cell with the same UMI.
The DataFrame holds the number of molecules in the group,
the sum of reads across all molecules in the group,
and the proportion of reads assigned to the most sequenced molecule.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Chimeric molecules are occasionally generated during library preparation for highly multiplexed droplet experiments.
Here, incomplete PCR products from one molecule hybridise to another molecule for extension using shared sequences like the poly-A tail for 3' protocols.
This produces an amplicon where the UMI and cell barcode originate from one transcript molecule but the gene sequence is from another.
If the second template is from another cell, this effect results in contamination of one cell's profile by another, similar to the contamination between samples discussed in <code>?<a href='swappedDrops.html'>swappedDrops</a></code>.</p>
<p>Chimerism manifests as molecules that have the same UMI sequence and cell barcode but are assigned to different genes.
To remove them, this function will simply discard all molecules within the same cell that share UMI sequences.
Of course, this may also remove non-chimeric molecules that have the same UMI by chance,
but for typical UMI lengths (10-12 bp for 10X protocols) we expect UMI collisions to be very rare between molecules from the same cell.</p>
<p>Nonetheless, to mitigate losses due to collisions,
we retain any molecule that has a much greater number of reads compared to all other molecules with the same UMI in the same cell.
This is based on the expectation that the original non-chimeric molecule will have undergone more rounds of PCR amplification compared to its chimeric offspring, and thus will have higher read coverage.
For all molecules with the same UMI within a given cell, we compute the proportion of reads assigned to each molecule and we keep the molecule with a proportion above <code>min.frac</code>.
If no molecule passes this threshold, the entire set is discarded.</p>
<p>The <code>use.library</code> argument can be used to only check for chimeras within a given feature type, e.g., CRISPR tags.
This is most relevant in situations where <code>sample</code> contains multiple libraries that involve different sets of shared sequences,
such that chimeras are unlikely to form between molecules from different libraries.
Analysis of just one library can be achieved by setting <code>use.library</code> to the name or index of the desired feature set.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Dixit A. (2016).
Correcting chimeric crosstalk in single cell RNA-seq experiments.
<em>biorXiv</em>, <a href='https://doi.org/10.1101/093237'>https://doi.org/10.1101/093237</a></p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Aaron Lun</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up some 10x HDF5-formatted data.</span>
<span class='va'>curfile</span> <span class='op'>&lt;-</span> <span class='fu'>DropletUtils</span><span class='fu'>:::</span><span class='fu'>simBasicMolInfo</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>out</span> <span class='op'>&lt;-</span> <span class='fu'>chimericDrops</span><span class='op'>(</span><span class='va'>curfile</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>out</span><span class='op'>$</span><span class='va'>cleaned</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1]  20 256</div><div class='input'>
<span class='va'>out2</span> <span class='op'>&lt;-</span> <span class='fu'>chimericDrops</span><span class='op'>(</span><span class='va'>curfile</span>, get.diagnostics<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>out2</span><span class='op'>$</span><span class='va'>diagnostics</span>
</div><div class='output co'>#&gt; DataFrame with 9534 rows and 3 columns
#&gt;         number     reads      prop
#&gt;      &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;
#&gt; 1            1        10         1
#&gt; 2            1        11         1
#&gt; 3            1        16         1
#&gt; 4            1         9         1
#&gt; 5            1        11         1
#&gt; ...        ...       ...       ...
#&gt; 9530         1        10         1
#&gt; 9531         1         9         1
#&gt; 9532         1        11         1
#&gt; 9533         1        11         1
#&gt; 9534         1        11         1</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


