<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>CellRanger's emptyDrops variant — emptyDropsCellRanger • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="CellRanger's emptyDrops variant — emptyDropsCellRanger" />
<meta property="og:description" content="An approximate implementation of the --soloCellFilter EmptyDrops_CR filtering approach, 
which itself was reverse-engineered from the behavior of CellRanger 3." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>CellRanger's emptyDrops variant</h1>
    
    <div class="hidden name"><code>emptyDropsCellRanger.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>An approximate implementation of the <code>--soloCellFilter EmptyDrops_CR</code> filtering approach, 
which itself was reverse-engineered from the behavior of CellRanger 3.</p>
    </div>

    <pre class="usage"><span class='fu'>emptyDropsCellRanger</span><span class='op'>(</span><span class='va'>m</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>emptyDropsCellRanger</span><span class='op'>(</span>
  <span class='va'>m</span>,
  n.expected.cells <span class='op'>=</span> <span class='fl'>3000</span>,
  max.percentile <span class='op'>=</span> <span class='fl'>0.99</span>,
  max.min.ratio <span class='op'>=</span> <span class='fl'>10</span>,
  umi.min <span class='op'>=</span> <span class='fl'>500</span>,
  umi.min.frac.median <span class='op'>=</span> <span class='fl'>0.01</span>,
  can.max.n <span class='op'>=</span> <span class='fl'>20000</span>,
  ind.min <span class='op'>=</span> <span class='fl'>45000</span>,
  ind.max <span class='op'>=</span> <span class='fl'>90000</span>,
  round <span class='op'>=</span> <span class='cn'>TRUE</span>,
  niters <span class='op'>=</span> <span class='fl'>10000</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>SerialParam</span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>emptyDropsCellRanger</span><span class='op'>(</span><span class='va'>m</span>, <span class='va'>...</span>, assay.type <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>m</th>
      <td><p>A numeric matrix-like object containing counts, where columns represent barcoded droplets and rows represent features.
The matrix should only contain barcodes for an individual sample, prior to any filtering for barcodes.</p></td>
    </tr>
    <tr>
      <th>n.expected.cells</th>
      <td><p>An integer scalar specifying the number of expected cells in a sample.</p></td>
    </tr>
    <tr>
      <th>max.percentile</th>
      <td><p>A numeric scalar between 0 and 1 used to define the maximum UMI count used in the simple filtering algorithm.</p></td>
    </tr>
    <tr>
      <th>max.min.ratio</th>
      <td><p>An integer scalar specifying the ratio of the maximum and minimum UMI count used in the simple filtering algorithm.</p></td>
    </tr>
    <tr>
      <th>umi.min</th>
      <td><p>An integer scalar specifying the minimum UMI count for inclusion of a barcode in the cell candidate pool.</p></td>
    </tr>
    <tr>
      <th>umi.min.frac.median</th>
      <td><p>A numeric scalar between 0 and 1 used to define the minimum UMI count for inclusion of a barcode in the cell candidate pool.</p></td>
    </tr>
    <tr>
      <th>ind.min</th>
      <td><p>An integer scalar specifying the lowest UMI count ranking for inclusion of a barcode in the ambient profile.</p></td>
    </tr>
    <tr>
      <th>ind.max</th>
      <td><p>An integer scalar specifying the highest UMI count ranking for inclusion of a barcode in the ambient profile.</p></td>
    </tr>
    <tr>
      <th>round</th>
      <td><p>A logical scalar indicating whether to check for non-integer values in <code>m</code> and, if present, round them for ambient profile estimation (see <code>?<a href='ambientProfileEmpty.html'>ambientProfileEmpty</a></code>) and the multinomial simulations.</p></td>
    </tr>
    <tr>
      <th>niters</th>
      <td><p>An integer scalar specifying the number of iterations to use for the Monte Carlo p-value calculations.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>A BiocParallelParam object indicating whether parallelization should be used.</p></td>
    </tr>
    <tr>
      <th>cand.max.n</th>
      <td><p>An integer scalar specifying the maximum number of barcodes that can be included in the cell candidate pool.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A DataFrame with the same fields as that returned by <code><a href='emptyDrops.html'>emptyDrops</a></code>.</p>
    <h2 class="hasAnchor" id="details-about-emptydropscellranger-arguments"><a class="anchor" href="#details-about-emptydropscellranger-arguments"></a>Details about <code>emptyDropsCellRanger</code> arguments</h2>

    

<p>The arguments in <span class="pkg">STARsolo</span> are one-to-one correspond to the arguments in <code>emptyDropsCellRanger</code>:</p><ul>
<li><p><code>n.expected.cells</code>: This argument is used to define the number of expected cells in a sample. 
It is the same as <code>nExpectedCells</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>max.percentile</code>: Together with <code>n.expected.cells</code>, <code>max.percentile</code> defines the maximum UMI count that an expected cell can have. 
This argument is same as <code>maxPercentile</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>max.min.ratio</code>: Based on the maximum UMI count computed using <code>n.expected.cells</code> and <code>max.percentile</code>, <code>max.min.ratio</code> defines the threshold of the minimum UMI count that an expected cell can have according to their ratio. 
The barcodes whose UMI count is above this threshold are assigned as real cells by the simple filtering algorithm (<code>nCellsSimple</code> in <span class="pkg">STARsolo</span>). 
It is same as <code>maxMinRatio</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>umi.min</code>: This argument is one of the three thresholds used to define the lower bound of the cell candidate pool.
<code>umi.min</code> specifies the minimum UMI count for inclusion of a barcode in the cell candidate pool. 
This argument is the same as <code>umiMin</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>umi.min.frac.median</code>: This argument is one of the three thresholds used to define the lower bound of the cell candidate pool. 
This argument defines the minimum UMI as <code>umi.min.frac.median</code> times the median UMI count of the real cells assigned by the simple filtering algorithm. 
This argument is the same as <code>umiMinFracMedian</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>cand.max.n</code>: This argument is one of the three thresholds used to define the lower bound of the cell candidate pool. 
This argument defines the minimum UMI as the UMI count of the barcodes whose UMI count ranks as the <code>cand.max.n</code>-th after the real cells defined by the simple filtering algorithm. 
This argument is the same as <code>umiMinFracMedian</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>ind.min</code>: This argument specifies the lowest UMI count ranking of the barcodes in the ambient pool. 
This argument is same as <code>indMin</code> in <span class="pkg">STARsolo</span>.</p></li>
<li><p><code>ind.max</code>: This argument specifies the highest UMI count ranking of the barcodes in the ambient pool. 
This argument is same as <code>indMin</code> in <span class="pkg">STARsolo</span>.</p></li>
</ul>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Kaminow et al. (2021).
STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus RNA-seq data
<a href='https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1'>https://www.biorxiv.org/content/10.1101/2021.05.05.442755v1</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='emptyDrops.html'>emptyDrops</a></code>, for the original implementation.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Dongze He, Rob Patro</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up some data:</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span>
<span class='va'>my.counts</span> <span class='op'>&lt;-</span> <span class='fu'>DropletUtils</span><span class='fu'>:::</span><span class='fu'>simCounts</span><span class='op'>(</span><span class='op'>)</span>

<span class='co'># Identify likely cell-containing droplets.</span>
<span class='va'>out</span> <span class='op'>&lt;-</span> <span class='fu'>emptyDropsCellRanger</span><span class='op'>(</span><span class='va'>my.counts</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in ambient.FUN(m, totals): no counts available to estimate the ambient profile</span></div><div class='input'><span class='va'>out</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'out' not found</span></div><div class='input'>
<span class='va'>is.cell</span> <span class='op'>&lt;-</span> <span class='va'>out</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;=</span> <span class='fl'>0.01</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'out' not found</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>is.cell</span>, na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'is.cell' not found</span></div><div class='input'>
<span class='co'># Subsetting the matrix to the cell-containing droplets.</span>
<span class='co'># (using 'which()' to handle NAs smoothly).</span>
<span class='va'>cell.counts</span> <span class='op'>&lt;-</span> <span class='va'>my.counts</span><span class='op'>[</span>,<span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>is.cell</span><span class='op'>)</span>,drop<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>]</span>
</div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'j' in selecting a method for function '[': error in evaluating the argument 'x' in selecting a method for function 'which': object 'is.cell' not found</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>cell.counts</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'cell.counts' not found</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


