<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Demultiplex cell hashing data — hashedDrops • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Demultiplex cell hashing data — hashedDrops" />
<meta property="og:description" content="Demultiplex cell barcodes into their samples of origin based on the most abundant hash tag oligo (HTO).
Also identify potential doublets based on the presence of multiple significant HTOs." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Demultiplex cell hashing data</h1>
    
    <div class="hidden name"><code>hashedDrops.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Demultiplex cell barcodes into their samples of origin based on the most abundant hash tag oligo (HTO).
Also identify potential doublets based on the presence of multiple significant HTOs.</p>
    </div>

    <pre class="usage"><span class='fu'>hashedDrops</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='co'># S4 method for ANY</span>
<span class='fu'>hashedDrops</span><span class='op'>(</span>
  <span class='va'>x</span>,
  ambient <span class='op'>=</span> <span class='cn'>NULL</span>,
  min.prop <span class='op'>=</span> <span class='fl'>0.05</span>,
  pseudo.count <span class='op'>=</span> <span class='fl'>5</span>,
  constant.ambient <span class='op'>=</span> <span class='cn'>FALSE</span>,
  doublet.nmads <span class='op'>=</span> <span class='fl'>3</span>,
  doublet.min <span class='op'>=</span> <span class='fl'>2</span>,
  doublet.mixture <span class='op'>=</span> <span class='cn'>FALSE</span>,
  confident.nmads <span class='op'>=</span> <span class='fl'>3</span>,
  confident.min <span class='op'>=</span> <span class='fl'>2</span>,
  combinations <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>hashedDrops</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span>, assay.type <span class='op'>=</span> <span class='st'>"counts"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A numeric/integer matrix-like object containing UMI counts.
Rows correspond to HTOs and columns correspond to cell barcodes.
Each barcode is assumed to correspond to a cell, i.e., cell calling is assumed to have already been performed.</p>
<p>Alternatively, a SummarizedExperiment object containing such a matrix.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>For the generic, further arguments to pass to individual methods.</p>
<p>For the SummarizedExperiment method, further arguments to pass to the ANY method.</p></td>
    </tr>
    <tr>
      <th>ambient</th>
      <td><p>A numeric vector of length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(x)</a></code>,
specifying the relative abundance of each HTO in the ambient solution - see Details.</p></td>
    </tr>
    <tr>
      <th>min.prop</th>
      <td><p>Numeric scalar to be used to infer the ambient profile when <code>ambient=NULL</code>, see <code><a href='ambientProfileBimodal.html'>ambientProfileBimodal</a></code>.</p></td>
    </tr>
    <tr>
      <th>pseudo.count</th>
      <td><p>A numeric scalar specifying the minimum pseudo-count when computing log-fold changes.</p></td>
    </tr>
    <tr>
      <th>constant.ambient</th>
      <td><p>Logical scalar indicating whether a constant level of ambient contamination should be used to estimate <code>LogFC2</code> for all cells.</p></td>
    </tr>
    <tr>
      <th>doublet.nmads</th>
      <td><p>A numeric scalar specifying the number of median absolute deviations (MADs) to use to identify doublets.</p></td>
    </tr>
    <tr>
      <th>doublet.min</th>
      <td><p>A numeric scalar specifying the minimum threshold on the log-fold change to use to identify doublets.</p></td>
    </tr>
    <tr>
      <th>doublet.mixture</th>
      <td><p>Logical scalar indicating whether to use a 2-component mixture model to identify doublets.</p></td>
    </tr>
    <tr>
      <th>confident.nmads</th>
      <td><p>A numeric scalar specifying the number of MADs to use to identify confidently assigned singlets.</p></td>
    </tr>
    <tr>
      <th>confident.min</th>
      <td><p>A numeric scalar specifying the minimum threshold on the log-fold change to use to identify singlets.</p></td>
    </tr>
    <tr>
      <th>combinations</th>
      <td><p>An integer matrix specifying valid <em>combinations</em> of HTOs.
Each row corresponds to a single sample and specifies the indices of rows in <code>x</code> corresponding to the HTOs used to label that sample.</p></td>
    </tr>
    <tr>
      <th>assay.type</th>
      <td><p>Integer or string specifying the assay containing the count matrix.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A DataFrame with one row per column of <code>x</code>, containing the following fields:</p><ul>
<li><p><code>Total</code>, integer specifying the total count across all HTOs for each barcode.</p></li>
<li><p><code>Best</code>, integer specifying the HTO with the highest abundance for each barcode.</p></li>
<li><p><code>Second</code>, integer specifying the HTO with the second-highest abundance.</p></li>
<li><p><code>LogFC</code>, numeric containing the log-fold change between the abundances of the best and second-best HTO.</p></li>
<li><p><code>LogFC2</code>, numeric containing the log-fold change in the second-best HTO over the ambient contamination.</p></li>
<li><p><code>Doublet</code>, logical specifying whether a barcode is a doublet.</p></li>
<li><p><code>Confident</code>, logical specifying whether a barcode is a confidently assigned singlet.</p></li>
</ul><p>In addition, the metadata contains <code>ambient</code>, a numeric vector containing the (estimate of the) ambient profile;
<code>doublet.threshold</code>, the threshold applied to <code>LogFC2</code> to identify doublets;
and <code>confident.threshold</code>, the threshold applied to non-doublet <code>LogFC</code> values to identify confident singlets.</p>
<p>If <code>combinations</code> is specified, <code>Best</code> instead specifies the sample (i.e., row index of <code>combinations</code>).
The interpretation of <code>LogFC</code> and <code>LogFC2</code> are slightly different, and <code>Second</code> is not reported - see &#8220;Resolving combinatorial hashes&#8221;.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The idea behind cell hashing is that cells from the same sample are stained with reagent conjugated with a single HTO.
Cells are mixed across multiple samples and subjected to droplet-based single-cell sequencing.
Cell barcode libraries can then be demultiplexed into individual samples based on whether their unique HTO is detected.</p>
<p>We identify the sample of origin for each cell barcode as that corresponding to the most abundant HTO.
(See below for more details on exactly how &#8220;most abundant&#8221; is defined.)
The log-fold change between the largest and second-largest abundances is reported for each barcode (<code>LogFC</code>), 
with large log-fold changes representing confident assignment to a single sample.
We also report the log-fold change of the second-most abundant HTO over the estimated level of ambient contamination (<code>LogFC2</code>),
with large log-fold changes indicating that a doublet is present.</p>
<p>To facilitate quality control, we explicitly identify problematic barcodes as outliers on the relevant metrics.</p><ul>
<li><p>By default, we identify putative doublets as those with <code>LogFC2</code> values that are
(i) <code>doublet.nmads</code> MADs above the median and (ii) greater than <code>doublet.min</code>.
The hard threshold is more-or-less arbitrary and aims to avoid overly aggressive detection 
of large outliers in a naturally right-skewed distribution 
(given that the log-fold changes are positive by definition, and most of the distribution is located near zero).</p></li>
<li><p>Alternatively, if <code>doublet.mixture=TRUE</code>, we fit a two-component mixture model to the <code>LogFC2</code> distribution.
Doublets are identified as all members of the component with the larger mean.
This avoids the need for the arbitrary parameters mentioned above but only works when there are many doublets,
otherwise both components will be fitted to the non-doublet values.
(Initialization of the model assumes at least 5% doublets.)</p></li>
</ul><p>Of the non-doublet libraries, we consider them to be confidently assigned to a single sample if their <code>LogFC</code> values are (i) <em>not</em> less than <code>confident.nmads</code> MADs below the median and (ii) greater than <code>confident.min</code>.
The hard threshold is again arbitrary, but this time it aims to avoid insufficiently aggressive outlier detection - 
typically from an inflation of the MAD when the <code>LogFC</code> values are large, positive and highly variable.</p>
    <h2 class="hasAnchor" id="use-only-on-non-empty-droplets"><a class="anchor" href="#use-only-on-non-empty-droplets"></a>Use only on non-empty droplets</h2>

    

<p>This function assumes that cell calling has already been performed, e.g., with <code><a href='emptyDrops.html'>emptyDrops</a></code>.
Specifically, <code>x</code> should only contain columns corresponding to non-empty droplets.
If empty droplets are included, their log-fold changes will simply reflect stochastic sampling in the ambient solution
and violate the assumptions involved in outlier detection.</p>
<p>If <code>x</code> contains columns for both empty and non-empty droplets,
it is straightforward to simply run <code><a href='emptyDrops.html'>emptyDrops</a></code> on the HTO count matrix to identify the latter.
Note that some fiddling with the <code>lower=</code> argument may be required,
depending on the sequencing depth of the HTO libraries.</p>
    <h2 class="hasAnchor" id="adjusting-abundances-for-ambient-contamination"><a class="anchor" href="#adjusting-abundances-for-ambient-contamination"></a>Adjusting abundances for ambient contamination</h2>

    

<p>HTO abundances require some care to compute due to the presence of ambient contamination in each library.
Ideally, the experiment would be performed in such a manner that the concentration of each HTO is the same.
However, if one HTO is present at higher concentration in the ambient solution,
this might incorrectly cause us to assign all barcodes to the corresponding sample.</p>
<p>To adjust for ambient contamination, we assume that the ambient contamination in each library follows the same profile as <code>ambient</code>.
We further assume that a minority of HTOs in a library are actually driven by the presence of cell(s), the rest coming from the ambient solution.
We estimate the level of ambient contamination in each barcode by scaling <code>ambient</code>, using a <span class="pkg">DESeq</span>-like normalization algorithm to compute the scaling factor.
(The requisite assumption of a non-DE majority follows from the two assumptions above.)
We then subtract the scaled ambient proportions from the HTO count profile to remove the effect of contamination.
Abundances that would otherwise be negative are set to zero.</p>
<p>The scaling factor for each barcode is defined by computing ratios between the HTO counts and <code>ambient</code>, and taking the median across all HTOs.
However, this strict definition is only used when there are at least 5 HTOs being considered.
For experiments with 3-4 HTOs, we assume that higher-order multiplets are negligible and define the scaling factor as the third-largest ratio.
For experiments with only 2 HTOs, the second-most abundant HTO is always used to estimate the ambient contamination.</p>
<p>Ideally, <code>ambient</code> would be obtained from libraries that do not correspond to cell-containing droplets.
For example, we could get this information from the <code>metadata</code> of the <code><a href='emptyDrops.html'>emptyDrops</a></code> output,
had we run <code><a href='emptyDrops.html'>emptyDrops</a></code> on the HTO count matrix (see below).
Unfortunately, in some cases (e.g., public data), counts are provided for only the cell-containing barcodes.
If <code>ambient=NULL</code>, the profile is inferred from <code>x</code> using <code><a href='ambientProfileBimodal.html'>ambientProfileBimodal</a></code>.</p>
    <h2 class="hasAnchor" id="computing-the-log-fold-changes"><a class="anchor" href="#computing-the-log-fold-changes"></a>Computing the log-fold changes</h2>

    

<p>HTO abundances may be set to zero after subtracting the ambient noise.
Thus, we need to add a pseudo-count to ensure that we can actually compute the log-fold changes described in &#8220;Value&#8221;.</p>
<p>For each barcode, we define the pseudo-count as the average ambient HTO count, i.e., the average of the scaled <code>ambient</code> for that barcode. 
This is motivated by the assumption that the number of contaminating transcript molecules is roughly the same in each droplet,
such that any differences in ambient coverage between libraries reflect barcode-specific biases (capture efficiency, sequencing depth) that would also affect cell-derived HTO counts. 
By using the average ambient count as the pseudo-count, we ensure that the shrinkage of the log-fold changes is not driven by the sequencing depth,
e.g., a constant pseudo-count would inflict greater shrinkage on libraries that have not been sequenced as deeply.
This avoids excessive variability in the log-fold change distribution that would otherwise reduce the precision of outlier detection.
Another nice aspect of this approach is that it collapses to a no-op if the experiment is well-executed with identical concentrations of all HTOs in the ambient solution.
(That said, we still enforce a minimum pseudo-count of <code>pseudo.count</code> if the average ambient count is lower than that,
simply to avoid highly variable log-fold changes when dealing with very low counts.)</p>
<p>Once the pseudo-count is added to the ambient-subtracted abundances, we compute the log-fold changes as described in &#8220;Value&#8221;.
<code>LogFC</code> is defined as the log-fold change in the most abundant HTO over the second-most abundant HTO.
<code>LogFC2</code> is defined as the log-fold change in the second-most abundant HTO over the ambient contamination.
By default, the denominator for <code>LogFC2</code> is set to the per-barcode average ambient count, equivalent to the pseudo-count used above.
This cancels out any variation in sequencing depth for more precise outlier calls.</p>
    <h2 class="hasAnchor" id="handling-or-fewer-samples"><a class="anchor" href="#handling-or-fewer-samples"></a>Handling 2 or fewer samples</h2>

    

<p>If <code>x</code> has no more than two rows, <code>LogFC2</code>, <code>Doublet</code> and <code>doublet.threshold</code> are set to <code>NA</code>.
Strictly speaking, doublet detection is not possible as the second HTO is always used to estimate the ambient scaling and thus <code>LogFC2</code> is always zero.
<code>Confident</code> calls are still available in the output of this function so assignment to the individual samples can still be performed.
In this scenario, the non-confident assignments are probably also doublets, though this cannot be said with much certainty.</p>
<p>To work around this limitation, we can set <code>constant.ambient=TRUE</code>,
which defines the denominator of each barcode's <code>LogFC2</code> as the median of the per-barcode average ambient counts across all barcodes.
This is useful in scenarios where <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(x)</a></code> is too small and we cannot assume that the abundances of most HTOs are driven by ambient contamination.
By assuming most barcodes are not doublets, we can obtain a dataset-wide baseline for the ambient contamination to compute <code>LogFC2</code>.
The cost of this approach is that the log-fold changes will be more variable as sequencing depth is not cancelled out.</p>
<p>If <code>x</code> has no more than one row, <code>Confident</code>, <code>LogFC</code> and <code>confident.threshold</code> are set to <code>NA</code>.
Obviously, if there is only one HTO, the identity of the assigned sample is a foregone conclusion.</p>
    <h2 class="hasAnchor" id="resolving-combinatorial-hashes"><a class="anchor" href="#resolving-combinatorial-hashes"></a>Resolving combinatorial hashes</h2>

    

<p>In some applications, samples are labelled with a combination of HTOs to enable achieve greater multiplexing throughput.
This is accommodated by passing <code>combinations</code> to specify the valid HTO combinations that were used for sample labelling.
Each row of <code>combinations</code> corresponds to a sample and should contain non-duplicated row indices of <code>x</code> corresponding to the HTOs used in that sample.</p>
<p>The calculation for the single-HTO case is then generalized for HTO combinations.
The most important differences are that:</p><ul>
<li><p>The reported <code>LogFC</code> is now the log-fold change between the \(n\)th most abundant HTO and the \(n+1\)th HTO,
where \(n\) is the number of HTOs in a valid combination. 
This captures the drop-off in abundance beyond the expected number of HTOs.</p></li>
<li><p>The reported <code>LogFC2</code> is now the log-fold change of the \(n+1\)th HTO over the ambient contamination.
This captures the high abundance of the more-than-expected number of HTOs when doublets are present.</p></li>
<li><p><code>Best</code> no longer refers to the row index of <code>x</code>, but instead to the row index of <code>combinations</code>.
This may contain <code>NA</code> values if a particular combination of HTOs is observed but not present in the expected set.</p></li>
<li><p><code>Second</code> is no longer reported as we cannot conveniently determine the identity of the second sample.</p></li>
</ul>

<p>We also generalize the edge-case behavior when there are not enough HTOs to support doublet detection. 
Consider that an inter-sample doublet may result in up to \(2n\) abundant HTOs.
Estimation of the scaling factor will attempt to avoid using the top \(2n\) ratios.
If <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(x)</a></code> is equal to or less than \(2n\), doublet statistics will not be reported at all, 
i.e., <code>Doublet</code> and <code>LogFC2</code> are set to <code>NA</code>.
This can be overcome by setting <code>constant.ambient=TRUE</code> as described above.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Stoeckius M, Zheng S, Houck-Loomis B et al. (2018)
Cell Hashing with barcoded antibodies enables multiplexing and doublet detection for single cell genomics.
<em>Genome Biol.</em> 19, 1:224</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='emptyDrops.html'>emptyDrops</a></code>, to identify which barcodes are likely to contain cells.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Aaron Lun</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up an example dataset with 10 HTOs and 10% doublets.</span>
<span class='va'>ncells</span> <span class='op'>&lt;-</span> <span class='fl'>1000</span>
<span class='va'>nhto</span> <span class='op'>&lt;-</span> <span class='fl'>10</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/Poisson.html'>rpois</a></span><span class='op'>(</span><span class='va'>ncells</span><span class='op'>*</span><span class='va'>nhto</span>, <span class='fl'>50</span><span class='op'>)</span>, nrow<span class='op'>=</span><span class='va'>nhto</span><span class='op'>)</span>
<span class='va'>true.sample</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='va'>nhto</span>, <span class='va'>ncells</span>, replace<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>y</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>true.sample</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='va'>ncells</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fl'>1000</span>

<span class='va'>ndoub</span> <span class='op'>&lt;-</span> <span class='va'>ncells</span><span class='op'>/</span><span class='fl'>10</span>
<span class='va'>next.sample</span> <span class='op'>&lt;-</span> <span class='op'>(</span><span class='va'>true.sample</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='va'>ndoub</span><span class='op'>]</span>  <span class='op'>+</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>%%</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span>
<span class='va'>next.sample</span><span class='op'>[</span><span class='va'>next.sample</span><span class='op'>==</span><span class='fl'>0</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span>
<span class='va'>y</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>next.sample</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='va'>ndoub</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fl'>500</span>

<span class='co'># Computing hashing statistics.</span>
<span class='va'>stats</span> <span class='op'>&lt;-</span> <span class='fu'>hashedDrops</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span>

<span class='co'># Doublets show up in the top-left, singlets in the bottom right.</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>stats</span><span class='op'>$</span><span class='va'>LogFC</span>, <span class='va'>stats</span><span class='op'>$</span><span class='va'>LogFC2</span><span class='op'>)</span>
</div><div class='img'><img src='hashedDrops-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Most cells should be singlets with low second log-fold changes.</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/hist.html'>hist</a></span><span class='op'>(</span><span class='va'>stats</span><span class='op'>$</span><span class='va'>LogFC2</span>, breaks<span class='op'>=</span><span class='fl'>50</span><span class='op'>)</span>
</div><div class='img'><img src='hashedDrops-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Identify confident singlets or doublets at the given threshold.</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>stats</span><span class='op'>$</span><span class='va'>Confident</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    Mode   FALSE    TRUE 
#&gt; logical     104     896 </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>stats</span><span class='op'>$</span><span class='va'>Doublet</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    Mode   FALSE    TRUE 
#&gt; logical     900     100 </div><div class='input'>
<span class='co'># Checking against the known truth, in this case</span>
<span class='co'># 'Best' contains the putative sample of origin.</span>
<span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='va'>stats</span><span class='op'>$</span><span class='va'>Best</span>, <span class='va'>true.sample</span><span class='op'>)</span> 
</div><div class='output co'>#&gt;     true.sample
#&gt;        1   2   3   4   5   6   7   8   9  10
#&gt;   1  114   0   0   0   0   0   0   0   0   0
#&gt;   2    0  93   0   0   0   0   0   0   0   0
#&gt;   3    0   0  97   0   0   0   0   0   0   0
#&gt;   4    0   0   0  96   0   0   0   0   0   0
#&gt;   5    0   0   0   0 110   0   0   0   0   0
#&gt;   6    0   0   0   0   0  95   0   0   0   0
#&gt;   7    0   0   0   0   0   0 102   0   0   0
#&gt;   8    0   0   0   0   0   0   0  76   0   0
#&gt;   9    0   0   0   0   0   0   0   0 112   0
#&gt;   10   0   0   0   0   0   0   0   0   0 105</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


