<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read the 10X molecule information file — read10xMolInfo • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Read the 10X molecule information file — read10xMolInfo" />
<meta property="og:description" content="Extract relevant fields from the molecule information HDF5 file, produced by CellRanger for 10X Genomics data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read the 10X molecule information file</h1>
    
    <div class="hidden name"><code>read10xMolInfo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract relevant fields from the molecule information HDF5 file, produced by CellRanger for 10X Genomics data.</p>
    </div>

    <pre class="usage"><span class='fu'>read10xMolInfo</span><span class='op'>(</span>
  <span class='va'>sample</span>,
  barcode.length <span class='op'>=</span> <span class='cn'>NULL</span>,
  keep.unmapped <span class='op'>=</span> <span class='cn'>FALSE</span>,
  get.cell <span class='op'>=</span> <span class='cn'>TRUE</span>,
  get.umi <span class='op'>=</span> <span class='cn'>TRUE</span>,
  get.gem <span class='op'>=</span> <span class='cn'>TRUE</span>,
  get.gene <span class='op'>=</span> <span class='cn'>TRUE</span>,
  get.reads <span class='op'>=</span> <span class='cn'>TRUE</span>,
  get.library <span class='op'>=</span> <span class='cn'>TRUE</span>,
  extract.library.info <span class='op'>=</span> <span class='cn'>FALSE</span>,
  version <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"auto"</span>, <span class='st'>"2"</span>, <span class='st'>"3"</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sample</th>
      <td><p>A string containing the path to the molecule information HDF5 file.</p></td>
    </tr>
    <tr>
      <th>barcode.length</th>
      <td><p>An integer scalar specifying the length of the cell barcode.
Only relevant when <code>version="2"</code>.</p></td>
    </tr>
    <tr>
      <th>keep.unmapped</th>
      <td><p>A logical scalar indicating whether unmapped molecules should be reported.</p></td>
    </tr>
    <tr>
      <th>get.cell, get.umi, get.gem, get.gene, get.reads, get.library</th>
      <td><p>Logical scalar indicating whether the corresponding field should be extracted for each molecule.</p></td>
    </tr>
    <tr>
      <th>extract.library.info</th>
      <td><p>Logical scalar indicating whether the library information should be extracted.
Only relevant when <code>version="3"</code>.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>String specifying the version of the 10X molecule information format to read data from.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list is returned containing <code>data</code>, 
a DataFrame where each row corresponds to a single transcript molecule.
This contains the following fields:</p><dl>
<dt><code>barcode</code>:</dt><dd><p>Character, the cell barcode for each molecule.</p></dd>
<dt><code>umi</code>:</dt><dd><p>Integer, the processed UMI barcode in 2-bit encoding.</p></dd> 
<dt><code>gem_group</code>:</dt><dd><p>Integer, the GEM group.</p></dd>
<dt><code>gene</code>:</dt><dd><p>Integer, the index of the gene to which the molecule was assigned.
This refers to an entry in the <code>genes</code> vector, see below.</p></dd>
<dt><code>reads</code>:</dt><dd><p>Integer, the number of reads mapped to this molecule.</p></dd>
<dt><code>reads</code>:</dt><dd><p>Integer, the number of reads mapped to this molecule.</p></dd>
<dt><code>library</code>:</dt><dd><p>Integer, the library index in cases where multiple libraries are present in the same file.
Only reported when <code>version="3"</code>.</p></dd>

</dl><p>A field will not be present in the DataFrame if the corresponding <code>get.*</code> argument is <code>FALSE</code>,</p>
<p>The second element of the list is <code>genes</code>, a character vector containing the names of all genes in the annotation.
This is indexed by the <code>gene</code> field in the <code>data</code> DataFrame.</p>
<p>If <code>version="3"</code>, a <code>feature.type</code> entry is added to the list.
This is a character vector of the same length as <code>genes</code>, containing the feature type for each gene.</p>
<p>If <code>extract.library.info=TRUE</code>, an additional element named <code>library.info</code> is returned.
This is a list of lists containing per-library information such as the <code>"library_type"</code>.
The <code>library</code> field in the <code>data</code> DataFrame indexes this list.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Molecules that were not assigned to any gene have <code>gene</code> set to <code>length(genes)+1</code>.
By default, these are removed when <code>keep.unmapped=FALSE</code>.</p>
<p>CellRanger 3.0 introduced a major change in the format of the molecule information files.
When <code>version="auto"</code>, the function will attempt to determine the version format of the file.
This can also be user-specified by setting <code>version</code> explicitly.</p>
<p>For files produced by version 2.2 of the CellRanger software, the length of the cell barcode is not given.
Instead, the barcode length is automatically inferred if <code>barcode.length=NULL</code> and <code>version="2"</code>.
Currently, version 1 of the 10X chemistry uses 14 nt barcodes, while version 2 uses 16 nt barcodes.</p>
<p>Setting any of the <code>get.*</code> arguments will (generally) avoid extraction of the corresponding field.
This can improve efficiency if that field is not necessary for further analysis.
Aside from the missing field, the results are guaranteed to be identical, i.e., same order and number of rows.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zheng GX, Terry JM, Belgrader P, and others (2017).
Massively parallel digital transcriptional profiling of single cells. 
<em>Nat Commun</em> 8:14049.</p>
<p>10X Genomics (2017).
Molecule info.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/molecule_info'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/molecule_info</a></p>
<p>10X Genomics (2018).
Molecule info.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/molecule_info'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/molecule_info</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='makeCountMatrix.html'>makeCountMatrix</a></code>, which creates a count matrix from this information.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Aaron Lun,
based on code by Jonathan Griffiths</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up some 10X HDF5-formatted data.</span>
<span class='va'>out</span> <span class='op'>&lt;-</span> <span class='fu'>DropletUtils</span><span class='fu'>:::</span><span class='fu'>simBasicMolInfo</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Reading the resulting file.</span>
<span class='fu'>read10xMolInfo</span><span class='op'>(</span><span class='va'>out</span><span class='op'>)</span>
</div><div class='output co'>#&gt; $data
#&gt; DataFrame with 9508 rows and 5 columns
#&gt;             cell       umi gem_group      gene     reads
#&gt;      &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt; 1           TAGG    742310         1        11        12
#&gt; 2           CAAG    492991         1         8         9
#&gt; 3           ACTG    823582         1        13         7
#&gt; 4           CCGT     37501         1         1        10
#&gt; 5           ACTT   1023713         1         5         7
#&gt; ...          ...       ...       ...       ...       ...
#&gt; 9504        TCTT    646811         1         6        11
#&gt; 9505        CCAG    229795         1        12         6
#&gt; 9506        GACT    616670         1         1         7
#&gt; 9507        ACGC    956761         1         6        12
#&gt; 9508        AGAA    121926         1         8        10
#&gt; 
#&gt; $genes
#&gt;  [1] "ENSG1"  "ENSG2"  "ENSG3"  "ENSG4"  "ENSG5"  "ENSG6"  "ENSG7"  "ENSG8" 
#&gt;  [9] "ENSG9"  "ENSG10" "ENSG11" "ENSG12" "ENSG13" "ENSG14" "ENSG15" "ENSG16"
#&gt; [17] "ENSG17" "ENSG18" "ENSG19" "ENSG20"
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


