<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load data from a 10X Genomics experiment — read10xCounts • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Load data from a 10X Genomics experiment — read10xCounts" />
<meta property="og:description" content="Creates a SingleCellExperiment from the CellRanger output directories for 10X Genomics data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load data from a 10X Genomics experiment</h1>
    
    <div class="hidden name"><code>read10xCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates a SingleCellExperiment from the CellRanger output directories for 10X Genomics data.</p>
    </div>

    <pre class="usage"><span class='fu'>read10xCounts</span><span class='op'>(</span>
  <span class='va'>samples</span>,
  sample.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>)</span>,
  col.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
  type <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"auto"</span>, <span class='st'>"sparse"</span>, <span class='st'>"HDF5"</span>, <span class='st'>"prefix"</span><span class='op'>)</span>,
  version <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"auto"</span>, <span class='st'>"2"</span>, <span class='st'>"3"</span><span class='op'>)</span>,
  genome <span class='op'>=</span> <span class='cn'>NULL</span>,
  compressed <span class='op'>=</span> <span class='cn'>NULL</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>SerialParam</span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>samples</th>
      <td><p>A character vector containing one or more directory names, each corresponding to a 10X sample.
Each directory should contain a matrix file, a gene/feature annotation file, and a barcode annotation file.</p>
<p>Alternatively, each string may contain a path to a HDF5 file in the sparse matrix format generated by 10X.
These can be mixed with directory names when <code>type="auto"</code>.</p>
<p>Alternatively, each string may contain a prefix of names for the three-file system described above,
where the rest of the name of each file follows the standard 10X output.</p></td>
    </tr>
    <tr>
      <th>sample.names</th>
      <td><p>A character vector of length equal to <code>samples</code>, containing the sample names to store in the column metadata of the output object.
If <code>NULL</code>, the file paths in <code>samples</code> are used directly.</p></td>
    </tr>
    <tr>
      <th>col.names</th>
      <td><p>A logical scalar indicating whether the columns of the output object should be named with the cell barcodes.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>String specifying the type of 10X format to read data from.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>String specifying the version of the 10X format to read data from.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>String specifying the genome if <code>type="HDF5"</code> and <code>version='2'</code>.</p></td>
    </tr>
    <tr>
      <th>compressed</th>
      <td><p>Logical scalar indicating whether the text files are compressed for <code>type="sparse"</code> or <code>"prefix"</code>.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>A BiocParallelParam object specifying how loading should be parallelized for multiple <code>samples</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A SingleCellExperiment object containing count data for each gene (row) and cell (column) across all <code>samples</code>.</p><ul>
<li><p>Row metadata will contain the fields <code>"ID"</code> and <code>"Symbol"</code>.
The former is the gene identifier (usually Ensembl), while the latter is the gene name.
If <code>version="3"</code>, it will also contain the <code>"Type"</code> field specifying the type of feature (e.g., gene or antibody).</p></li>
<li><p>Column metadata will contain the fields <code>"Sample"</code> and <code>"Barcode"</code>.
The former contains the name of the sample (or if not supplied, the path in <code>samples</code>) from which each column was obtained.
The latter contains to the cell barcode sequence and GEM group for each cell library.</p></li>
<li><p>Rows are named with the gene identifier.
Columns are named with the cell barcode in certain settings, see Details.</p></li>
<li><p>The assays will contain a single <code>"counts"</code> matrix, containing UMI counts for each gene in each cell.
Note that the matrix representation will depend on the format of the <code>samples</code>, see Details.</p></li>
<li><p>The metadata contains a <code>"Samples"</code> field, containing the input <code>samples</code> character vector.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function has a long and storied past.
It was originally developed as the <code>read10xResults</code> function in <span class="pkg">scater</span>, inspired by the <code>Read10X</code> function from the <span class="pkg">Seurat</span> package.
It was then migrated to this package in an effort to consolidate some 10X-related functionality across various packages.</p>
<p>If <code>type="auto"</code>, the format of the input file is automatically detected for each <code>samples</code> based on whether it ends with <code>".h5"</code>.
If so, <code>type</code> is set to <code>"HDF5"</code>; otherwise it is set to <code>"sparse"</code>.</p><ul>
<li><p>If <code>type="sparse"</code>, count data are loaded as a dgCMatrix object.
This is a conventional column-sparse compressed matrix format produced by the CellRanger pipeline,
consisting of a (possibly Gzipped) MatrixMarket text file (<code>"matrix.mtx"</code>)
with additional tab-delimited files for barcodes (<code>"barcodes.tsv"</code>)
and gene annotation (<code>"features.tsv"</code> for version 3 or <code>"genes.tsv"</code> for version 2).</p></li>
<li><p>If <code>type="prefix"</code>, count data are also loaded as a dgCMatrix object.
This assumes the same three-file structure for each sample as described for <code>type="sparse"</code>,
but each sample is defined here by a prefix in the file names rather than by being a separate directory.
For example, if the <code>samples</code> entry is <code>"xyx_"</code>,
the files are expected to be <code>"xyz_matrix.mtx"</code>, <code>"xyz_barcodes.tsv"</code>, etc.</p></li>
<li><p>If <code>type="HDF5"</code>, count data are assumed to follow the 10X sparse HDF5 format for large data sets.
It is loaded as a TENxMatrix object, which is a stub object that refers back to the file in <code>samples</code>.
Users may need to set <code>genome</code> if it cannot be automatically determined when <code>version="2"</code>.</p></li>
</ul>

<p>When <code>type="sparse"</code> or <code>"prefix"</code> and <code>compressed=NULL</code>,
the function will automatically search for both the unzipped and Gzipped versions of the files.
This assumes that the compressed files have an additional <code>".gz"</code> suffix.
We can restrict to only compressed or uncompressed files by setting <code>compressed=TRUE</code> or <code>FALSE</code>, respectively.</p>
<p>CellRanger 3.0 introduced a major change in the format of the output files for both <code>type</code>s.
If <code>version="auto"</code>, the version of the format is automatically detected from the supplied paths.
For <code>type="sparse"</code>, this is based on whether there is a <code>"features.tsv.gz"</code> file in the directory.
For <code>type="HDF5"</code>, this is based on whether there is a top-level <code>"matrix"</code> group with a <code>"matrix/features"</code> subgroup in the file.</p>
<p>Matrices are combined by column if multiple <code>samples</code> were specified.
This will throw an error if the gene information is not consistent across <code>samples</code>.</p>
<p>If <code>col.names=TRUE</code> and <code>length(sample)==1</code>, each column is named by the cell barcode.
For multiple samples, the index of each sample in <code>samples</code> is concatenated to the cell barcode to form the column name.
This avoids problems with multiple instances of the same cell barcodes in different samples.</p>
<p>Note that user-level manipulation of sparse matrices requires loading of the <span class="pkg">Matrix</span> package.
Otherwise, calculation of <code>rowSums</code>, <code>colSums</code>, etc. will result in errors.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zheng GX, Terry JM, Belgrader P, and others (2017).
Massively parallel digital transcriptional profiling of single cells. 
<em>Nat Commun</em> 8:14049.</p>
<p>10X Genomics (2017).
Gene-Barcode Matrices.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/matrices</a></p>
<p>10X Genomics (2018).
Feature-Barcode Matrices.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices</a></p>
<p>10X Genomics (2018).
HDF5 Gene-Barcode Matrix Format.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/advanced/h5_matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/advanced/h5_matrices</a></p>
<p>10X Genomics (2018).
HDF5 Feature Barcode Matrix Format.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>splitAltExps</code>, to split alternative feature sets (e.g., antibody tags) into their own Experiments.</p>
<p><code><a href='write10xCounts.html'>write10xCounts</a></code>, to create 10X-formatted file(s) from a count matrix.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Davis McCarthy, with modifications from Aaron Lun</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up some 10X genomics output.</span>
<span class='fu'><a href='https://rdrr.io/r/utils/example.html'>example</a></span><span class='op'>(</span><span class='va'>write10xCounts</span><span class='op'>)</span>
</div><div class='output co'>#&gt; 
#&gt; wrt10C&gt; # Mocking up some count data.
#&gt; wrt10C&gt; library(Matrix)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'Matrix'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:S4Vectors':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand</span></div><div class='output co'>#&gt; 
#&gt; wrt10C&gt; my.counts &lt;- matrix(rpois(1000, lambda=5), ncol=10, nrow=100)
#&gt; 
#&gt; wrt10C&gt; my.counts &lt;- as(my.counts, "dgCMatrix")
#&gt; 
#&gt; wrt10C&gt; cell.ids &lt;- paste0("BARCODE-", seq_len(ncol(my.counts)))
#&gt; 
#&gt; wrt10C&gt; ngenes &lt;- nrow(my.counts)
#&gt; 
#&gt; wrt10C&gt; gene.ids &lt;- paste0("ENSG0000", seq_len(ngenes))
#&gt; 
#&gt; wrt10C&gt; gene.symb &lt;- paste0("GENE", seq_len(ngenes))
#&gt; 
#&gt; wrt10C&gt; # Writing this to file:
#&gt; wrt10C&gt; tmpdir &lt;- tempfile()
#&gt; 
#&gt; wrt10C&gt; write10xCounts(tmpdir, my.counts, gene.id=gene.ids, 
#&gt; wrt10C+     gene.symbol=gene.symb, barcodes=cell.ids)
#&gt; 
#&gt; wrt10C&gt; list.files(tmpdir)
#&gt; [1] "barcodes.tsv" "genes.tsv"    "matrix.mtx"  
#&gt; 
#&gt; wrt10C&gt; # Creating a version 3 HDF5 file:
#&gt; wrt10C&gt; tmph5 &lt;- tempfile(fileext=".h5")
#&gt; 
#&gt; wrt10C&gt; write10xCounts(tmph5, my.counts, gene.id=gene.ids, 
#&gt; wrt10C+     gene.symbol=gene.symb, barcodes=cell.ids, version='3')</div><div class='input'>
<span class='co'># Reading it in.</span>
<span class='va'>sce10x</span> <span class='op'>&lt;-</span> <span class='fu'>read10xCounts</span><span class='op'>(</span><span class='va'>tmpdir</span><span class='op'>)</span>

<span class='co'># Column names are dropped with multiple 'samples'.</span>
<span class='va'>sce10x2</span> <span class='op'>&lt;-</span> <span class='fu'>read10xCounts</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>tmpdir</span>, <span class='va'>tmpdir</span><span class='op'>)</span><span class='op'>)</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


