<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Write count data in the 10x format — write10xCounts • DropletUtils</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Write count data in the 10x format — write10xCounts" />
<meta property="og:description" content="Create a directory containing the count matrix and cell/gene annotation from a sparse matrix of UMI counts, 
in the format produced by the CellRanger software suite." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletUtils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletUtils.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Write count data in the 10x format</h1>
    
    <div class="hidden name"><code>write10xCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create a directory containing the count matrix and cell/gene annotation from a sparse matrix of UMI counts, 
in the format produced by the CellRanger software suite.</p>
    </div>

    <pre class="usage"><span class='fu'>write10xCounts</span><span class='op'>(</span>
  <span class='va'>path</span>,
  <span class='va'>x</span>,
  barcodes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>,
  gene.id <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>,
  gene.symbol <span class='op'>=</span> <span class='va'>gene.id</span>,
  gene.type <span class='op'>=</span> <span class='st'>"Gene Expression"</span>,
  overwrite <span class='op'>=</span> <span class='cn'>FALSE</span>,
  type <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"auto"</span>, <span class='st'>"sparse"</span>, <span class='st'>"HDF5"</span><span class='op'>)</span>,
  genome <span class='op'>=</span> <span class='st'>"unknown"</span>,
  version <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"2"</span>, <span class='st'>"3"</span><span class='op'>)</span>,
  chemistry <span class='op'>=</span> <span class='st'>"Single Cell 3' v3"</span>,
  original.gem.groups <span class='op'>=</span> <span class='fl'>1L</span>,
  library.ids <span class='op'>=</span> <span class='st'>"custom"</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>path</th>
      <td><p>A string containing the path to the output directory (for <code>type="sparse"</code>) or file (for <code>type="HDF5"</code>).</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>A sparse numeric matrix of UMI counts.</p></td>
    </tr>
    <tr>
      <th>barcodes</th>
      <td><p>A character vector of cell barcodes, one per column of <code>x</code>.</p></td>
    </tr>
    <tr>
      <th>gene.id</th>
      <td><p>A character vector of gene identifiers, one per row of <code>x</code>.</p></td>
    </tr>
    <tr>
      <th>gene.symbol</th>
      <td><p>A character vector of gene symbols, one per row of <code>x</code>.</p></td>
    </tr>
    <tr>
      <th>gene.type</th>
      <td><p>A character vector of gene types, expanded to one per row of <code>x</code>.
Only used when <code>version="3"</code>.</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>A logical scalar specifying whether <code>path</code> should be overwritten if it already exists.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>String specifying the type of 10X format to save <code>x</code> to.
This is either a directory containing a sparse matrix with row/column annotation (<code>"sparse"</code>)
or a HDF5 file containing the same information (<code>"HDF5"</code>).</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>String specifying the genome for storage when <code>type="HDF5"</code>.
This can be a character vector with one genome per feature if <code>version="3"</code>.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>String specifying the version of the CellRanger format to produce.</p></td>
    </tr>
    <tr>
      <th>chemistry, original.gem.groups, library.ids</th>
      <td><p>Strings containing metadata attributes to be added to the HDF5 file for <code>type="HDF5"</code>.
Their interpretation is not formally documented and is left to the user's imagination.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For <code>type="sparse"</code>, a directory is produced at <code>path</code>.
If <code>version="2"</code>, this will contain the files <code>"matrix.mtx"</code>, <code>"barcodes.tsv"</code> and <code>"genes.tsv"</code>.
If <code>version="3"</code>, it will instead contain <code>"matrix.mtx.gz"</code>, <code>"barcodes.tsv.gz"</code> and <code>"features.tsv.gz"</code>.</p>
<p>For <code>type="HDF5"</code>, a HDF5 file is produced at <code>path</code> containing data in column-sparse format.
If <code>version="2"</code>, data are stored in the HDF5 group named <code>genome</code>.
If <code>version="3"</code>, data are stored in the group <code>"matrix"</code>.</p>
<p>A <code>TRUE</code> value is invisibly returned.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function will try to automatically detect the desired format based on whether <code>path</code> ends with <code>".h5"</code>.
If so, it assumes that <code>path</code> specifies a HDF5 file path and sets <code>type="HDF5"</code>.
Otherwise it will set <code>type="sparse"</code> under the assumption that <code>path</code> specifies a path to a directory.</p>
<p>Note that there were major changes in the output format for CellRanger version 3.0 to account for non-gene features such as antibody or CRISPR tags. 
Users can switch to this new format using <code>version="3"</code>.
See the documentation for &#8220;latest&#8221; for this new format, otherwise see &#8220;2.2&#8221; or earlier.</p>
<p>The primary purpose of this function is to create files to use for testing <code><a href='read10xCounts.html'>read10xCounts</a></code>.
In principle, it is possible to re-use the HDF5 matrices in <code>cellranger reanalyze</code>.
We recommend against doing so routinely due to CellRanger's dependence on undocumented metadata attributes that may change without notice.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>10X Genomics (2017).
Gene-Barcode Matrices.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/matrices</a></p>
<p>10X Genomics (2018).
Feature-Barcode Matrices.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices</a></p>
<p>10X Genomics (2018).
HDF5 Gene-Barcode Matrix Format.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/advanced/h5_matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/advanced/h5_matrices</a></p>
<p>10X Genomics (2018).
HDF5 Feature Barcode Matrix Format.
<a href='https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices'>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='read10xCounts.html'>read10xCounts</a></code>, to read CellRanger matrices into R.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Aaron Lun</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Mocking up some count data.</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://Matrix.R-forge.R-project.org/'>Matrix</a></span><span class='op'>)</span>
<span class='va'>my.counts</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/Poisson.html'>rpois</a></span><span class='op'>(</span><span class='fl'>1000</span>, lambda<span class='op'>=</span><span class='fl'>5</span><span class='op'>)</span>, ncol<span class='op'>=</span><span class='fl'>10</span>, nrow<span class='op'>=</span><span class='fl'>100</span><span class='op'>)</span>
<span class='va'>my.counts</span> <span class='op'>&lt;-</span> <span class='fu'>as</span><span class='op'>(</span><span class='va'>my.counts</span>, <span class='st'>"dgCMatrix"</span><span class='op'>)</span>
<span class='va'>cell.ids</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"BARCODE-"</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>my.counts</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>ngenes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>my.counts</span><span class='op'>)</span>
<span class='va'>gene.ids</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"ENSG0000"</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='va'>ngenes</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>gene.symb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"GENE"</span>, <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span><span class='op'>(</span><span class='va'>ngenes</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># Writing this to file:</span>
<span class='va'>tmpdir</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='fu'>write10xCounts</span><span class='op'>(</span><span class='va'>tmpdir</span>, <span class='va'>my.counts</span>, gene.id<span class='op'>=</span><span class='va'>gene.ids</span>, 
    gene.symbol<span class='op'>=</span><span class='va'>gene.symb</span>, barcodes<span class='op'>=</span><span class='va'>cell.ids</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/list.files.html'>list.files</a></span><span class='op'>(</span><span class='va'>tmpdir</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "barcodes.tsv" "genes.tsv"    "matrix.mtx"  </div><div class='input'>
<span class='co'># Creating a version 3 HDF5 file:</span>
<span class='va'>tmph5</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span>fileext<span class='op'>=</span><span class='st'>".h5"</span><span class='op'>)</span>
<span class='fu'>write10xCounts</span><span class='op'>(</span><span class='va'>tmph5</span>, <span class='va'>my.counts</span>, gene.id<span class='op'>=</span><span class='va'>gene.ids</span>, 
    gene.symbol<span class='op'>=</span><span class='va'>gene.symb</span>, barcodes<span class='op'>=</span><span class='va'>cell.ids</span>, version<span class='op'>=</span><span class='st'>'3'</span><span class='op'>)</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


